### ISSUE TITLE
fill struct fill wrong position

### ISSUE THREAD

**Issue by IWANABETHATGUY** [rust-analyzer/rust-analyzer#4314]
## main.rs
```rust
use hello::SelectBox;
use hello::{Button, Screen};

fn main() {
    let screen = Screen {
        components: vec![
            Box::new(Button {
                width: 20,
                height: 30,
            }),
            Box::new(SelectBox {})
        ],
    };
    screen.run();
}
```
## lib.rs
```rust
pub trait Draw {
    fn draw(&self);
}
pub trait Color {
    fn color(&self);
}
pub struct Screen {
    pub components: Vec<Box<dyn Draw>>,
}

impl Screen {
    pub fn run(&self) {
        for c in self.components.iter() {
            c.draw();
        }
    }
}

pub struct Button {
    pub width: usize,
    pub height: usize,
}

impl Draw for Button {
    fn draw(&self) {
        println!("this a  button",);
    }
}

impl Color for Button {
    fn color(&self) {
        println!("start paint the button ",)
    }
}

pub struct SelectBox {
    pub height: usize,
    pub width: usize,
}

impl Draw for SelectBox {
    fn draw(&self) {
        println!("this is a selectBox",)
    }
}

impl Color for SelectBox {
    fn color(&self) {
        println!("start paint the selectBox",)
    }
}
```
when i click code action `fill struct`
![image](https://user-images.githubusercontent.com/17974631/81048235-86042280-8eee-11ea-98ce-0eb757f2e267.png)
![image](https://user-images.githubusercontent.com/17974631/81048425-e5fac900-8eee-11ea-9469-5b50b2155eb0.png)
the result is above ,

**Comment by IWANABETHATGUY** [comment]
It seems that the code action has fill the wrong position

**Comment by edwin0cheng** [comment]
Note that `SelectBox` is inside a `vec![]` macro, seem like that we have some bugs for `fill-struct` inside macro.

**Comment by matklad** [comment]
Yup, we fail to remap the range of diangostic. The code in question is here:

https://github.com/rust-analyzer/rust-analyzer/blob/b211c5814e994a411095e4392d804d692d58d43b/crates/ra_ide/src/diagnostics.rs#L84-L99

The simple fix would be to just not show quick-fixes in macros. I wonder, however, how the real fix should look like here? 

We *can* apply the fix to macro expanded code and then remap text edit back, but that seems wrong -- the fix *should* take whitespace into account. For example, in IntelliJ we used a rule like "if there will be more than three fields after the fix, also place each existing field on a separate line". 

I think what we need here is to reparse **original** source file as recrod literal. That is, we get a fix for a struct literal inside a macro. We map the range of that literal upwords and get a range in the original file. That range contains a token tree naturally. We then **reparse** the raw text of the range as record literal, and calculate the fix on that.

**Comment by matklad** [comment]
No updates -- we have minimal examples, pointers to the code which handles diagnostics, and a couple of plans for fixing outlined, but no one is actively working on a fix.
